src/utils/Cypher.ts

```diff
@@ -26,19 +32,79 @@ export default class Cypher {
     return this.configuration;
   }
 
-  public static encrypt(data: string): string {
+  public static async detectConfigurationKey(tenantID: string): Promise<void> {
+
+    const configCryptoKey: string = Configuration.getCryptoConfig().key;
+    const keySettings = await SettingStorage.getCryptoKeySettings(tenantID);
+    let cryptoKeySetting: CryptoKeySetting;
+
+    // Check if Crypto Key settings exists
+    if (keySettings) {
+      // Detect new config Cypher key
+      if (keySettings.cryptoKey.newKey && keySettings.cryptoKey.newKey !== configCryptoKey) {
+        cryptoKeySetting = {
+          oldKey: keySettings.cryptoKey.newKey
+        } as CryptoKeySetting;
+      } else if (!keySettings.cryptoKey.newKey && keySettings.cryptoKey.oldKey !== configCryptoKey) {
+        // If no newKey exist, detect oldKey change
+        cryptoKeySetting = {
+          oldKey: keySettings.cryptoKey.oldKey
+        } as CryptoKeySetting;
+      }
+      // If key change detected, prepare KeySettings change
+      if (cryptoKeySetting) {
+        await SettingStorage.deleteSetting(tenantID, keySettings.id);
+        cryptoKeySetting.newKey = configCryptoKey;
+
+        // Key migration of senitive data
+        this.cryptoKey = cryptoKeySetting;
+      }
+    } else {
+      // Create New Config Crypto Key in Tenant Settings
+      cryptoKeySetting = {
+        oldKey: configCryptoKey
+      } as CryptoKeySetting;
+    }
+
+    if (cryptoKeySetting) {
+      const keySettingToSave = {
+        identifier: TenantComponents.CRYPTO_KEY,
+        type: KeySettingsType.CRYPTO_KEY,
+        cryptoKey: cryptoKeySetting
+      } as KeySettings;
+      await SettingStorage.saveCryptoKeySettings(tenantID, keySettingToSave);
+    }
+  }
+
+  public static async migrateSensitiveData(tenantID: string, setting: SettingDB): Promise<void> {
+    if (this.cryptoKey) {
+      this.decryptSensitiveDataInJSON(setting, this.cryptoKey.oldKey);
+      this.encryptSensitiveDataInJSON(setting, this.cryptoKey.newKey);
+      await SettingStorage.saveSettings(tenantID, setting);
+    }
```

Comment:
I would really prefer to also have a code path in that method that test if the ciphered value with the key changed really match the 'clear' value deciphered with the formerKey for safety.

Reply:
@stefania-santimbrean unless I've missed something, you've implemented that review comment.